packages/salesforcedx-vscode-replay-debugger/src/adapter/apexReplayDebug.ts

```diff
@@ -26,21 +29,37 @@ import {
 import { LogContext } from '../core/logContext';
 import { nls } from '../messages';
 
+const TRACE_ALL = 'all';
+const TRACE_CATEGORY_PROTOCOL = 'protocol';
+const TRACE_CATEGORY_LOGFILE = 'logfile';
+const TRACE_CATEGORY_LAUNCH = 'launch';
+const TRACE_CATEGORY_BREAKPOINTS = 'breakpoints';
+
+export type TraceCategory =
+  | 'all'
+  | 'protocol'
+  | 'logfile'
+  | 'launch'
+  | 'breakpoints';
+
 export interface LaunchRequestArguments
   extends DebugProtocol.LaunchRequestArguments {
   logFile: string;
-  stopOnEntry: boolean | true;
-  trace: boolean | true;
+  stopOnEntry?: boolean | true;
+  trace?: boolean | string;
 }
 
-export class ApexReplayDebug extends DebugSession {
+export class ApexReplayDebug extends LoggingDebugSession {
   public static THREAD_ID = 1;
   protected logContext: LogContext;
+  protected trace: string[] = [];
+  protected traceAll = false;
   private breakpointUtil: BreakpointUtil;
   private initializedResponse: DebugProtocol.InitializeResponse;
+  protected breakpoints: Map<string, number[]> = new Map();
```

Comment:
Do we have any way to let the user know if the source and the debug logs have deviated?
With the Apex interactive debugger, we could detect if the user is trying to save to the server and terminate the debugger session.
With the Apex replay debugger, it's possible that an accidental <new line> could shift all the breakpoints such that nothing hits.
I don't have a solution off the top of my head but I'm wondering if there is anything that we can/should do here.

Reply:
For the apex interactive debugger, we only kill the session if they push their updated source. It was also possible for them to edit apex source locally, not push the source, and confuse the breakpoints. We have a filewatcher that detects changes to .cls/.trigger files and print out a warning if they change a source file while the session is active. We can make it more strict and terminate the session if we detect a local change.